Failed to parse http response

Hello,

I'm having an issue with handling error responses. A following ApolloParseException is being thrown:

`
Failed to parse http response
                                                            com.apollographql.apollo.exception.ApolloParseException: Failed to parse http response
                                                                at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor.parse(ApolloParseInterceptor.java:91)
                                                                at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor.intercept(ApolloParseInterceptor.java:48)
                                                                at com.apollographql.apollo.internal.interceptor.RealApolloInterceptorChain.proceed(RealApolloInterceptorChain.java:41)
                                                                at com.apollographql.apollo.internal.interceptor.ApolloCacheInterceptor.intercept(ApolloCacheInterceptor.java:84)
                                                                at com.apollographql.apollo.internal.interceptor.RealApolloInterceptorChain.proceed(RealApolloInterceptorChain.java:41)
                                                                at com.apollographql.apollo.internal.RealApolloCall.execute(RealApolloCall.java:128)
                                                                at com.apollographql.apollo.rx.RxApollo$2.call(RxApollo.java:96)
                                                                at com.apollographql.apollo.rx.RxApollo$2.call(RxApollo.java:92)
                                                                at rx.Single.subscribe(Single.java:1967)
                                                                at rx.Single$13$1.call(Single.java:2039)
                                                                at rx.internal.schedulers.CachedThreadScheduler$EventLoopWorker$1.call(CachedThreadScheduler.java:230)
                                                                at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55)
                                                                at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
                                                                at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                                at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:272)
                                                                at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
                                                                at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
                                                                at java.lang.Thread.run(Thread.java:762)
                                                             Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String java.lang.Object.toString()' on a null object reference
                                                                at com.apollographql.apollo.internal.interceptor.HttpResponseBodyParser.readError(HttpResponseBodyParser.java:99)
                                                                at com.apollographql.apollo.internal.interceptor.HttpResponseBodyParser.access$300(HttpResponseBodyParser.java:26)
                                                                at com.apollographql.apollo.internal.interceptor.HttpResponseBodyParser$2$1.read(HttpResponseBodyParser.java:86)
                                                                at com.apollographql.apollo.internal.interceptor.HttpResponseBodyParser$2$1.read(HttpResponseBodyParser.java:84)
                                                                at com.apollographql.apollo.internal.json.ResponseJsonStreamReader.nextObject(ResponseJsonStreamReader.java:86)
                                                                at com.apollographql.apollo.internal.interceptor.HttpResponseBodyParser$2.read(HttpResponseBodyParser.java:84)
                                                                at com.apollographql.apollo.internal.interceptor.HttpResponseBodyParser$2.read(HttpResponseBodyParser.java:82)
                                                                at com.apollographql.apollo.internal.json.ResponseJsonStreamReader.nextList(ResponseJsonStreamReader.java:101)
                                                                at com.apollographql.apollo.internal.interceptor.HttpResponseBodyParser.readResponseErrors(HttpResponseBodyParser.java:82)
                                                                at com.apollographql.apollo.internal.interceptor.HttpResponseBodyParser.parse(HttpResponseBodyParser.java:63)
                                                                at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor.parse(ApolloParseInterceptor.java:80)
                                                                at com.apollographql.apollo.internal.interceptor.ApolloParseInterceptor.intercept(ApolloParseInterceptor.java:48) 
                                                                at com.apollographql.apollo.internal.interceptor.RealApolloInterceptorChain.proceed(RealApolloInterceptorChain.java:41) 
                                                                at com.apollographql.apollo.internal.interceptor.ApolloCacheInterceptor.intercept(ApolloCacheInterceptor.java:84) 
                                                                at com.apollographql.apollo.internal.interceptor.RealApolloInterceptorChain.proceed(RealApolloInterceptorChain.java:41) 
                                                                at com.apollographql.apollo.internal.RealApolloCall.execute(RealApolloCall.java:128) 
                                                                at com.apollographql.apollo.rx.RxApollo$2.call(RxApollo.java:96) 
                                                                at com.apollographql.apollo.rx.RxApollo$2.call(RxApollo.java:92) 
                                                                at rx.Single.subscribe(Single.java:1967) 
                                                                at rx.Single$13$1.call(Single.java:2039) 
                                                                at rx.internal.schedulers.CachedThreadScheduler$EventLoopWorker$1.call(CachedThreadScheduler.java:230) 
                                                                at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55) 
                                                                at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428) 
                                                                at java.util.concurrent.FutureTask.run(FutureTask.java:237) 
                                                                at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:272) 
                                                                at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
                                                                at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
                                                                at java.lang.Thread.run(Thread.java:762) 
`

And this is the response causing the issue:
`{"data":null,"errors":[{"message":null,"code":"userNotFound","path":"loginWithPassword"}]}`

Seems like `readError(ResponseJsonStreamReader)` in `HttpResponseBodyParser.java` is missing null handling for this scenario.

Using 0.3.2-SNAPSHOT version of library and rx-support.

--- a/apollo-api/src/main/java/com/apollographql/apollo/api/Error.java
+++ b/apollo-api/src/main/java/com/apollographql/apollo/api/Error.java
@@ -18,7 +18,8 @@ public final class Error {
   private final List<Location> locations;
   private final Map<String, Object> customAttributes;
 
-  public Error(String message, @Nullable List<Location> locations, @Nullable Map<String, Object> customAttributes) {
+  public Error(@Nullable String message, @Nullable List<Location> locations,
+      @Nullable Map<String, Object> customAttributes) {
     this.message = message;
     this.locations = locations != null ? unmodifiableList(locations) : Collections.<Location>emptyList();
     this.customAttributes = customAttributes != null ? unmodifiableMap(customAttributes)
@@ -28,7 +29,7 @@ public final class Error {
   /**
    * Returns server error message.
    */
-  public String message() {
+  @Nullable public String message() {
     return message;
   }
 

--- a/apollo-integration/src/test/java/com/apollographql/apollo/IntegrationTest.java
+++ b/apollo-integration/src/test/java/com/apollographql/apollo/IntegrationTest.java
@@ -144,7 +144,7 @@ public class IntegrationTest {
         .asList("Gary Kurtz", "Rick McCallum"));
   }
 
-  @Test public void errorResponse() throws Exception {
+  @Test public void error_response() throws Exception {
     server.enqueue(mockResponse("ResponseError.json"));
     Response<AllPlanets.Data> body = apolloClient.query(new AllPlanets()).execute();
     assertThat(body.hasErrors()).isTrue();
@@ -152,6 +152,18 @@ public class IntegrationTest {
     assertThat(body.errors()).containsExactly(new Error(
         "Cannot query field \"names\" on type \"Species\".",
         Collections.singletonList(new Error.Location(3, 5)), Collections.<String, Object>emptyMap()));
+  }
+
+  @Test public void error_response_with_nulls_and_custom_attributes() throws Exception {
+    server.enqueue(mockResponse("ResponseErrorWithNullsAndCustomAttributes.json"));
+    Response<AllPlanets.Data> body = apolloClient.query(new AllPlanets()).execute();
+    assertThat(body.hasErrors()).isTrue();
+    assertThat(body.errors()).hasSize(1);
+    assertThat(body.errors().get(0).message()).isNull();
+    assertThat(body.errors().get(0).customAttributes()).hasSize(2);
+    assertThat(body.errors().get(0).customAttributes().get("code")).isEqualTo("userNotFound");
+    assertThat(body.errors().get(0).customAttributes().get("path")).isEqualTo("loginWithPassword");
+    assertThat(body.errors().get(0).locations()).hasSize(0);
   }
 
   @Test public void errorResponse_custom_attributes() throws Exception {

--- a/apollo-integration/src/test/resources/ResponseErrorWithNullsAndCustomAttributes.json
+++ b/apollo-integration/src/test/resources/ResponseErrorWithNullsAndCustomAttributes.json
@@ -1,0 +1,10 @@
+{
+  "errors": [
+    {
+      "message": null,
+      "locations": null,
+      "code":"userNotFound",
+      "path":"loginWithPassword"
+    }
+  ]
+}

--- a/apollo-runtime/src/main/java/com/apollographql/apollo/internal/interceptor/HttpResponseBodyParser.java
+++ b/apollo-runtime/src/main/java/com/apollographql/apollo/internal/interceptor/HttpResponseBodyParser.java
@@ -96,7 +96,8 @@ final class HttpResponseBodyParser<D extends Operation.Data, W> {
     final Map<String, Object> customAttributes = new HashMap<>();
     for (Map.Entry<String, Object> entry : reader.toMap().entrySet()) {
       if ("message".equals(entry.getKey())) {
-        message = entry.getValue().toString();
+        Object value = entry.getValue();
+        message = value != null ? value.toString() : null;
       } else if ("locations".equals(entry.getKey())) {
         List<Map<String, Object>> locationItems = (List<Map<String, Object>>) entry.getValue();
         if (locationItems != null) {

